python/hail/methods/statgen.py

```diff
@@ -1721,25 +1723,111 @@ def pc_relate(dataset, k, maf, block_size=512, min_kinship=-float("inf"), statis
+    maf : :obj:`float`
+        The minimum individual-specific allele frequency for an allele used to
+        measure relatedness.
+    path : :obj:`str` or :obj:`None`
+        A temporary directory to store intermediate matrices. Storing the matrices
+        to a file system is necessary for reliable execution of this method.
+    block_size : :obj:`int`
+        the side length of the blocks of the block-distributed matrices; this
+        should be set such that at least three of these matrices fit in memory
+        (in addition to all other objects necessary for Spark and Hail).
+    min_kinship : :obj:`float`
+        Pairs of samples with kinship lower than ``min_kinship`` are excluded
+        from the results.
+    statistics : :obj:`str`
+        the set of statistics to compute, `phi` will only compute the
+        kinship statistic, `phik2` will compute the kinship and
+        identity-by-descent two statistics, `phik2k0` will compute the
+        kinship statistics and both identity-by-descent two and zero, `all`
+        computes the kinship statistic and all three identity-by-descent
+        statistics.
 
+    Returns
+    -------
+    :class:`.Table`
+        A :class:`.Table` mapping pairs of samples to estimations of their
+        kinship and identity-by-descent zero, one, and two.
+
+        The fields of the resulting :class:`.Table` entries are of types: `i`:
+        ``ds.col_key.dtype``, `j`: ``ds.col_key.dtype``, `kin`:
+        :py:data:`.tfloat64`, `k2`: :py:data:`tfloat64`, `k1`:
+        :py:data:`tfloat64`, `k0`: :py:data:`tfloat64`. The table is keyed by
+        `i` and `j`.
     """
-    require_col_key_str(dataset, 'pc_relate')
-    dataset = require_biallelic(dataset, 'pc_relate')
-    intstatistics = {"phi": 0, "phik2": 1, "phik2k0": 2, "all": 3}[statistics]
-    _, scores, _ = hwe_normalized_pca(dataset, k, False, True)
+    assert(scores._indices.source == ds)
+    # analyze('pc_relate_with_scores', scores, ds._col_indices)
+    ds = ds.annotate_cols(scores=scores)
+    ds = require_biallelic(ds, 'pc_relate_with_scores')
+    ds = require_unique_samples(ds, 'pc_relate_with_scores')
+    ds = ds.select_entries(GT=ds.GT,
+                           naa=hl.float64(ds.GT.n_alt_alleles()))
+    ds = ds.select_rows(
+        *ds.row_key,
+        mean_gt=(agg.sum(ds.naa) /
+                 agg.count_where(hl.is_defined(ds.GT))))
+    mean_imputed_gt = hl.or_else(ds.naa, ds.mean_gt)
+    g = BlockMatrix.write_from_entry_expr(mean_imputed_gt,
+                                          path=path,
+                                          block_size=block_size)
+
+    pc_scores = (ds.add_col_index('column_index').cols().collect())
+    pc_scores.sort(key=lambda x: x.column_index)
+    pc_scores = [x.scores for x in pc_scores]
+
+    int_statistics = {"phi": 0, "phik2": 1, "phik2k0": 2, "all": 3}[statistics]
+
+    col_keys_table = ds.col_keys()
+    col_keys_local = col_keys_table.collect()
```

Comment:
Same with this - since you've already `collect()`ed the pc_scores, wouldn't it be a bit faster to just grab them from there rather than collect a second time?

Reply:
yes this would be better, cc: @jbloom22